### ISSUE TITLE
AErender doesn't exit when finished rendering...

### ISSUE THREAD

**Issue by Marcuzzz** [devtud/aerender#3]
I think it's has something to do with a plugin (Saber) but after running aerender it keeps running doing nothing after this:  
PROGRESS:  Total Time Elapsed: 1 Min, 2 Sec 
So I found your aerender wrapper which is helpfull but my knowledge is limited with asyncio. 
I tried a couple things:

```
#in this function async def check_and_raise(log_line: str):
#and also here in the async for log_line, log_tag in stdout_log_generator

if 'Total Time Elapsed' in log_line:     
    process_task.stop()
    # 
    process_task.cancel()
    #
    process_task.done()
```

it stops but very messy and after rendering I would like to some other stuff...
Is there away to stop nicely?

Example out after stopping with stop()
```
DEBUG:asyncio:taking long time to close proactor
Traceback (most recent call last):
  File "test.py", line 19, in <module>
    asyncio.run(coro)
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\base_events.py", line 584, in run_until_complete
    return future.result()
  File "C:\Users\marc.otten\Documents\mmogon-v2\venv\lib\site-packages\aerender\wrapper.py", line 112, in run
    result = await self.run_command(command)
  File "C:\Users\marc.otten\Documents\mmogon-v2\venv\lib\site-packages\aerender\wrapper.py", line 160, in run_command
    await check_and_raise(log_line)
  File "C:\Users\marc.otten\Documents\mmogon-v2\venv\lib\site-packages\aerender\wrapper.py", line 141, in check_and_raise
    process_task.stop()
AttributeError: '_asyncio.Task' object has no attribute 'stop'
Exception ignored in: <function BaseSubprocessTransport.__del__ at 0x000001FFB3E69048>
Traceback (most recent call last):
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\base_subprocess.py", line 127, in __del__
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\base_subprocess.py", line 104, in close
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\proactor_events.py", line 86, in close
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\base_events.py", line 688, in call_soon
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\base_events.py", line 480, in _check_closed
RuntimeError: Event loop is closed
Exception ignored in: <function _ProactorBasePipeTransport.__del__ at 0x000001FFB3E6B400>
Traceback (most recent call last):
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\proactor_events.py", line 95, in __del__
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\proactor_events.py", line 86, in close
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\base_events.py", line 688, in call_soon
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\base_events.py", line 480, in _check_closed
RuntimeError: Event loop is closed
```

**Comment by devtud** [comment]
Hi @Marcuzzz. Not sure I understand the problem very well. Are you trying to avoid using asyncio? If so, I'm afraid you won't be able, but even if your code didn't use asyncio, you can use it only for aerender. Unfortunatelly I'm not able to test, but in theory the example in [README.md](https://github.com/devtud/aerender/blob/master/README.md) should work.

In more details, you have to make the imports, then set the loop policy if you're using Windows (all in the main file).

```python
# imports ...

asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())
```

Then, you should be able to run the actual rendering where you need:

```python
AERENDER_FULLPATH = Path('C:/Program Files/Adobe/Adobe After Effects CC 2019/Support Files/aerender')
aerender = AERenderWrapper(exe_path=AERENDER_FULLPATH)
coro = aerender.run(
        project_path=Path('C:/projects/proj1.aep'),
        comp_name='Comp 1',
        output_path=Path('C:/output/proj1/proj1.avi'),
)
asyncio.run(coro)  # This line is blocking, so after this line your rendering is finished
```

You can provide your snippet of code where you want to integrate the rendering, maybe I can help more.

**Comment by Marcuzzz** [comment]
Hi,

Thanks for your quick response. 
This is the snippet:

wrapper.py:

```
import asyncio
import logging
from asyncio import StreamReader
from pathlib import Path

from .exceptions import (AERenderError, AfterEffectsError, CompositionNotFoundError,
                         PathNotFoundError)


__all__ = ['AERenderWrapper']

logging.basicConfig(level=logging.DEBUG)

logger = logging.getLogger('aerender')


class AERenderWrapper:
    def __init__(self, exe_path: Path):
        self.exe_path: Path = exe_path

    async def run(
        self, reuse: bool = None, project_path: Path = None, project_name: str = None,
        comp_name: str = None, index_in_render_queue: str = None,
        render_settings_template: str = None, output_module_template: str = None,
        output_path: Path = None, logfile_path: Path = None, start_frame: int = None,
        end_frame: int = None, increment: int = None, image_cache_percent: int = None,
        max_mem_percent: int = None, verbose_flag: str = None, close_flag: str = None,
        sound_flag: str = None, continue_on_missing_footage: bool = None,
        version: bool = None):
        """
        Args:
            reuse: use this flag if you want to try and reuse an already running instance
                of AE to perform the render. By default, aerender will launch a new
                instance of After Effects, even if one is already running. But, if AE is
                already running, and the -reuse flag is provided, then aerender will ask
                the already running instance of AE to perform the render. Whenever
                aerender launches a new instance of AE, it will tell AE to quit when
                rendering is completed; otherwise, it will not quit AE. Also, the
                preferences will be written to file upon quit when the -reuse flag is
                specified; otherwise it will not be written.
            project_path: a file path or URI specifying a project file to open. If none is
                provided, aerender will work with the currently open project. If no
                project is open and no project is provided, an error will result.
            project_name: a name of a team project to open
            comp_name: specifies a comp to be rendered. If the comp is in the render queue
                already, and in a queueable state, then (only) the first queueable
                instance of that comp on the render queue will be rendered. If the comp
                is in the project but not in the render queue, then it will be added to
                the render queue and rendered. If no -comp argument is provided,
                aerender will render the entire render queue as is. In this case
                (no -comp), the only other arguments used will be -project, -log, -v,
                -mem_usage, and -close; the -RStemplate, -OMtemplate, -output, -s,
                -e, and -i arguments will be ignored.
            index_in_render_queue: specifies a render queue item to be rendered. Options
                that make sense when rendering a single render queue item are available
                like with the -comp flag.
            render_settings_template: the name of a template to apply to the render
                queue item.If the template does not exist it is an error. Default is to
                use the render template already defined for the item
            output_module_template: the name of a template to apply to the output module.
                If the template does not exist it is an error. Default is to use the
                template already defined for the output module
            output_path: is a file path or URI specifying the destination render file.
                Default is the path already in the project file
            logfile_path: is a file path or URI specifying the location of the log file.
                Default is stdout
            start_frame: the first frame to render. Default is the start frame in the file
            end_frame: the last frame to render. Note, this is "inclusive;" the final
                frame will be rendered. Default is the end frame in the file
            increment: the number of frames to advance before rendering a new frame. A
                value of 1 (the default) results in a normal rendering of all frames.
                Higher increments will repeat the same (frame increment-1) times and then
                render a new one, starting the cycle again. Higher values result in
                faster renders but choppier motion. Default is 1
            image_cache_percent: specifies the maximum percent of memory used to cache
                already rendered images/footage
            max_mem_percent: specifies the total percent of memory that can be used
                by After Effects
            verbose_flag: specifies the type of messages reported. Possible values are
                ERRORS (prints only fatal and problem errors) or
                ERRORS_AND_PROGRESS (prints progress of rendering as well).
                Default value is ERRORS_AND_PROGRESS
            close_flag: specifies whether or not to close the project when done rendering,
                and whether or not to save changes. If close_flag is
                DO_NOT_SAVE_CHANGES, project will be closed without saving changes. If
                close_flag is SAVE_CHANGES, project will be closed and changes will
                be saved. If close_flag is DO_NOT_CLOSE the project will be left open;
                but the project is left open only if using an already-running instance
                of AE, since new invocations of AE must always close and quit when done.
                Default value is DO_NOT_SAVE_CHANGES
            sound_flag: specifies whether or not to play a sound when rendering is
                complete. Possible values are "ON" or "OFF". Default value is "OFF"
            continue_on_missing_footage: Do not stop rendering on missing footage. Log and
                render with placeholder color bars.
            version: displays the version number of aerender to the console. Does not
                render

        Returns:

        """
        command = await self.get_command(
            aerender_abs_path=self.exe_path, reuse=reuse, project_path=project_path,
            project_name=project_name, comp_name=comp_name,
            index_in_render_queue=index_in_render_queue,
            render_settings_template=render_settings_template,
            output_module_template=output_module_template, output_path=output_path,
            logfile_path=logfile_path, start_frame=start_frame, end_frame=end_frame,
            increment=increment, image_cache_percent=image_cache_percent,
            max_mem_percent=max_mem_percent, verbose_flag=verbose_flag,
            close_flag=close_flag, sound_flag=sound_flag,
            continue_on_missing_footage=continue_on_missing_footage, version=version)
        result = await self.run_command(command)
        return result

    @staticmethod
    async def run_command(command: str):
        async def log_generator(reader: StreamReader, tag: str):
            while True:
                line = await reader.readline()
                if not line:
                    break
                decoded = line.decode().strip()
                yield decoded, tag

        async def check_and_raise(log_line: str):
            if 'After Effects error' in log_line:
                if 'Path is not valid' in log_line:
                    raise PathNotFoundError(log_line)
                raise AfterEffectsError(log_line)
            if 'aerender ERROR' in log_line:
                if 'No comp was found with the given name' in log_line:
                    raise CompositionNotFoundError(log_line)
                raise AERenderError(log_line)
            if 'LoadLibrary "n" failed' in log_line:
                raise AERenderError(log_line)

        process = await asyncio.create_subprocess_shell(
            command,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE
        )

        process_task = asyncio.create_task(process.wait())

        #task = asyncio.gather(process_task)

        stdout_log_generator = log_generator(process.stdout, 'stdout')

        # TODO check and parse stderr also
        # stderr_log_generator = log_generator(process.stderr, 'stderr')

        async for log_line, log_tag in stdout_log_generator:
            logger.info(f'{log_tag}: {log_line}')
            #trying to stop the subprocess somehow without errors:
            if 'Finished composition' in log_line:
                print('Hi')
                try:
                    #process_task.cancel()
                    process_task.done()
                    #process_task.terminate()
                    #process_task.kill()
                    #process_task.stop()

                except:
                    sys.exit()
                    pass
                #return None
            #
            # if 'Total Time Elapsed' in log_line:
            #     process.terminate()
            #     return None

            # try:
            #     await check_and_raise(log_line)
            #
            # except AERenderError as e:
            #     logger.error(f'Waiting for process to finish due to error: {e}')
            #     await process_task
            #     raise e
            # except:
            #     pass

        #await asyncio.gather(process_task)

        #logger.info(f'[{command!r} exited with {process.returncode}]')

        #return None

    @staticmethod
    async def get_command(
        aerender_abs_path: Path, *, reuse: bool = None, project_path: Path = None,
        project_name: str = None, comp_name: str = None,
        index_in_render_queue: str = None, render_settings_template: str = None,
        output_module_template: str = None, output_path: Path = None,
        logfile_path: Path = None, start_frame: int = None, end_frame: int = None,
        increment: int = None, image_cache_percent: int = None,
        max_mem_percent: int = None, verbose_flag: str = None, close_flag: str = None,
        sound_flag: str = None, continue_on_missing_footage: bool = None,
        version: bool = None
    ) -> str:

        command = f'"{aerender_abs_path}"'
        if reuse:
            command += f' -reuse'
        if project_path is not None:
            command += f' -project "{project_path}"'
        if project_name is not None:
            command += f' -teamproject "{project_name}"'
        if comp_name is not None:
            command += f' -comp "{comp_name}"'
        if index_in_render_queue is not None:
            command += f' -rqindex {index_in_render_queue}'
        if render_settings_template is not None:
            command += f' -RStemplate "{render_settings_template}"'
        if output_module_template is not None:
            command += f' -OMtemplate "{output_module_template}"'
        if output_path is not None:
            command += f' -output "{output_path}"'
        if logfile_path is not None:
            command += f' -log "{logfile_path}"'
        if start_frame is not None:
            command += f' -s {start_frame}'
        if end_frame is not None:
            command += f' -e {end_frame}'
        if increment is not None:
            command += f' -i {increment}'
        if image_cache_percent is not None and max_mem_percent is not None:
            command += f' -mem_usage {image_cache_percent} {max_mem_percent}'
        if verbose_flag is not None:
            assert verbose_flag in ['ERRORS', 'ERRORS_AND_PROGRESS']
            command += f' -v {verbose_flag}'
        if close_flag is not None:
            assert close_flag in ['DO_NOT_SAVE_CHANGES', 'SAVE_CHANGES', 'DO_NOT_CLOSE']
            command += f' -close {close_flag}'
        if sound_flag is not None:
            assert sound_flag in ['ON', 'OFF']
            command += f' -sound {sound_flag}'
        if continue_on_missing_footage:
            command += f' -continueOnMissingFootage'
        if version:
            command += f' -version'
        return command

```

**Comment by devtud** [comment]
First of all I would like to understand what stops you from using the wrapper as is. Why exactly are you trying to modify it? What exactly does not work when you just use the wrapper as is in your app?

**Comment by Marcuzzz** [comment]
Yes of course. So AErender doesn't stop... It hangs after the render is finished:
<img width="1055" alt="Screenshot 2020-09-11 at 10 23 01" src="https://user-images.githubusercontent.com/7437046/92892209-d9482d80-f418-11ea-9771-38246e566655.png">

Never experienced this before!

**Comment by devtud** [comment]
Could you also provide the a minimal code snippet that reproduces your problem? Thanks!

**Comment by Marcuzzz** [comment]
Your code is fine can't reproduce it with your code... 
It's AErender that doesn't stop. It might be something with the particular project... I've cleaned it for testing and it seems to stop after. 

But for the sake and for any future occasion I still would like to be able to
Stop the subprocess nicely when:
`if 'Finished composition' in log_line` 
or
`if 'Total Time Elapsed' in log_line`

Is this possible?
Maybe with a timeout of 5sec...

**Comment by devtud** [comment]
Oh, ok. Maybe you can stop the process when it logs what you want, something like:

```pyhton
if 'Finished composition' in log_line:
    process_task.kill()
```

and receive errors on `asyncio.gather` with `return_exceptions=True` like:

```
result = await asyncio.gather(process_task, return_exceptions=True)
```

and handle the exception in `result`. But I cannot test this so I'm not sure if it helps.

**Comment by Marcuzzz** [comment]
Changed wrapper.py to this:

(added some comments what I've tried)

```
import asyncio
import logging
from asyncio import StreamReader
from pathlib import Path

from .exceptions import (AERenderError, AfterEffectsError, CompositionNotFoundError,
                         PathNotFoundError)


__all__ = ['AERenderWrapper']

logging.basicConfig(level=logging.DEBUG)

logger = logging.getLogger('aerender')


class AERenderWrapper:
    def __init__(self, exe_path: Path):
        self.exe_path: Path = exe_path

    async def run(
        self, reuse: bool = None, project_path: Path = None, project_name: str = None,
        comp_name: str = None, index_in_render_queue: str = None,
        render_settings_template: str = None, output_module_template: str = None,
        output_path: Path = None, logfile_path: Path = None, start_frame: int = None,
        end_frame: int = None, increment: int = None, image_cache_percent: int = None,
        max_mem_percent: int = None, verbose_flag: str = None, close_flag: str = None,
        sound_flag: str = None, continue_on_missing_footage: bool = None,
        version: bool = None):
        """
        Args:
            reuse: use this flag if you want to try and reuse an already running instance
                of AE to perform the render. By default, aerender will launch a new
                instance of After Effects, even if one is already running. But, if AE is
                already running, and the -reuse flag is provided, then aerender will ask
                the already running instance of AE to perform the render. Whenever
                aerender launches a new instance of AE, it will tell AE to quit when
                rendering is completed; otherwise, it will not quit AE. Also, the
                preferences will be written to file upon quit when the -reuse flag is
                specified; otherwise it will not be written.
            project_path: a file path or URI specifying a project file to open. If none is
                provided, aerender will work with the currently open project. If no
                project is open and no project is provided, an error will result.
            project_name: a name of a team project to open
            comp_name: specifies a comp to be rendered. If the comp is in the render queue
                already, and in a queueable state, then (only) the first queueable
                instance of that comp on the render queue will be rendered. If the comp
                is in the project but not in the render queue, then it will be added to
                the render queue and rendered. If no -comp argument is provided,
                aerender will render the entire render queue as is. In this case
                (no -comp), the only other arguments used will be -project, -log, -v,
                -mem_usage, and -close; the -RStemplate, -OMtemplate, -output, -s,
                -e, and -i arguments will be ignored.
            index_in_render_queue: specifies a render queue item to be rendered. Options
                that make sense when rendering a single render queue item are available
                like with the -comp flag.
            render_settings_template: the name of a template to apply to the render
                queue item.If the template does not exist it is an error. Default is to
                use the render template already defined for the item
            output_module_template: the name of a template to apply to the output module.
                If the template does not exist it is an error. Default is to use the
                template already defined for the output module
            output_path: is a file path or URI specifying the destination render file.
                Default is the path already in the project file
            logfile_path: is a file path or URI specifying the location of the log file.
                Default is stdout
            start_frame: the first frame to render. Default is the start frame in the file
            end_frame: the last frame to render. Note, this is "inclusive;" the final
                frame will be rendered. Default is the end frame in the file
            increment: the number of frames to advance before rendering a new frame. A
                value of 1 (the default) results in a normal rendering of all frames.
                Higher increments will repeat the same (frame increment-1) times and then
                render a new one, starting the cycle again. Higher values result in
                faster renders but choppier motion. Default is 1
            image_cache_percent: specifies the maximum percent of memory used to cache
                already rendered images/footage
            max_mem_percent: specifies the total percent of memory that can be used
                by After Effects
            verbose_flag: specifies the type of messages reported. Possible values are
                ERRORS (prints only fatal and problem errors) or
                ERRORS_AND_PROGRESS (prints progress of rendering as well).
                Default value is ERRORS_AND_PROGRESS
            close_flag: specifies whether or not to close the project when done rendering,
                and whether or not to save changes. If close_flag is
                DO_NOT_SAVE_CHANGES, project will be closed without saving changes. If
                close_flag is SAVE_CHANGES, project will be closed and changes will
                be saved. If close_flag is DO_NOT_CLOSE the project will be left open;
                but the project is left open only if using an already-running instance
                of AE, since new invocations of AE must always close and quit when done.
                Default value is DO_NOT_SAVE_CHANGES
            sound_flag: specifies whether or not to play a sound when rendering is
                complete. Possible values are "ON" or "OFF". Default value is "OFF"
            continue_on_missing_footage: Do not stop rendering on missing footage. Log and
                render with placeholder color bars.
            version: displays the version number of aerender to the console. Does not
                render

        Returns:

        """
        command = await self.get_command(
            aerender_abs_path=self.exe_path, reuse=reuse, project_path=project_path,
            project_name=project_name, comp_name=comp_name,
            index_in_render_queue=index_in_render_queue,
            render_settings_template=render_settings_template,
            output_module_template=output_module_template, output_path=output_path,
            logfile_path=logfile_path, start_frame=start_frame, end_frame=end_frame,
            increment=increment, image_cache_percent=image_cache_percent,
            max_mem_percent=max_mem_percent, verbose_flag=verbose_flag,
            close_flag=close_flag, sound_flag=sound_flag,
            continue_on_missing_footage=continue_on_missing_footage, version=version)
        result = await self.run_command(command)
        return result

    @staticmethod
    async def run_command(command: str):
        async def log_generator(reader: StreamReader, tag: str):
            while True:
                line = await reader.readline()
                if not line:
                    break
                decoded = line.decode().strip()
                yield decoded, tag

        async def check_and_raise(log_line: str):
            if 'After Effects error' in log_line:
                if 'Path is not valid' in log_line:
                    raise PathNotFoundError(log_line)
                raise AfterEffectsError(log_line)
            if 'aerender ERROR' in log_line:
                if 'No comp was found with the given name' in log_line:
                    raise CompositionNotFoundError(log_line)
                raise AERenderError(log_line)
            if 'LoadLibrary "n" failed' in log_line:
                raise AERenderError(log_line)
            
            #also tried
            #if 'Finished composition' in log_line:
                # await self.process_task.kill()
                # also tried:
                # await process_task.kill()
                # process_task.kill()


        process = await asyncio.create_subprocess_shell(
            command,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE
        )

        process_task = asyncio.create_task(process.wait())

        #task = asyncio.gather(process_task)

        stdout_log_generator = log_generator(process.stdout, 'stdout')

        # TODO check and parse stderr also
        # stderr_log_generator = log_generator(process.stderr, 'stderr')

        async for log_line, log_tag in stdout_log_generator:
            logger.info(f'{log_tag}: {log_line}')

            if 'Finished composition' in log_line:
                await self.process_task.kill()
                # also tried:
                # await process_task.kill()
                # process_task.kill()

            try:
                await check_and_raise(log_line)

            except AERenderError as e:
                logger.error(f'Waiting for process to finish due to error: {e}')
                await process_task
                raise e
            except:
                pass



        result = await asyncio.gather(process_task, return_exceptions=True)

        logger.info(f'[{command!r} exited with {process.returncode}]')

        return result

    @staticmethod
    async def get_command(
        aerender_abs_path: Path, *, reuse: bool = None, project_path: Path = None,
        project_name: str = None, comp_name: str = None,
        index_in_render_queue: str = None, render_settings_template: str = None,
        output_module_template: str = None, output_path: Path = None,
        logfile_path: Path = None, start_frame: int = None, end_frame: int = None,
        increment: int = None, image_cache_percent: int = None,
        max_mem_percent: int = None, verbose_flag: str = None, close_flag: str = None,
        sound_flag: str = None, continue_on_missing_footage: bool = None,
        version: bool = None
    ) -> str:

        command = f'"{aerender_abs_path}"'
        if reuse:
            command += f' -reuse'
        if project_path is not None:
            command += f' -project "{project_path}"'
        if project_name is not None:
            command += f' -teamproject "{project_name}"'
        if comp_name is not None:
            command += f' -comp "{comp_name}"'
        if index_in_render_queue is not None:
            command += f' -rqindex {index_in_render_queue}'
        if render_settings_template is not None:
            command += f' -RStemplate "{render_settings_template}"'
        if output_module_template is not None:
            command += f' -OMtemplate "{output_module_template}"'
        if output_path is not None:
            command += f' -output "{output_path}"'
        if logfile_path is not None:
            command += f' -log "{logfile_path}"'
        if start_frame is not None:
            command += f' -s {start_frame}'
        if end_frame is not None:
            command += f' -e {end_frame}'
        if increment is not None:
            command += f' -i {increment}'
        if image_cache_percent is not None and max_mem_percent is not None:
            command += f' -mem_usage {image_cache_percent} {max_mem_percent}'
        if verbose_flag is not None:
            assert verbose_flag in ['ERRORS', 'ERRORS_AND_PROGRESS']
            command += f' -v {verbose_flag}'
        if close_flag is not None:
            assert close_flag in ['DO_NOT_SAVE_CHANGES', 'SAVE_CHANGES', 'DO_NOT_CLOSE']
            command += f' -close {close_flag}'
        if sound_flag is not None:
            assert sound_flag in ['ON', 'OFF']
            command += f' -sound {sound_flag}'
        if continue_on_missing_footage:
            command += f' -continueOnMissingFootage'
        if version:
            command += f' -version'
        return command


```

This is how it now stops:

```
INFO:aerender:stdout: PROGRESS:  11/09/2020 15:48:31: Finished composition "DP_HERO_ARTIST_ALLCAPS_AM_LOGO_NOREVEAL_BLUE_1x1_optimized_v001".
DEBUG:asyncio:taking long time to close proactor
Stopped!
Stopped!!!!
Exception ignored in: <function BaseSubprocessTransport.__del__ at 0x000001E59E983F28>
Traceback (most recent call last):
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\base_subprocess.py", line 127, in __del__
    self.close()
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\base_subprocess.py", line 104, in close
    proto.pipe.close()
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\proactor_events.py", line 86, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\base_events.py", line 688, in call_soon
    self._check_closed()
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\base_events.py", line 480, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function _ProactorBasePipeTransport.__del__ at 0x000001E59E9AD378>
Traceback (most recent call last):
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\proactor_events.py", line 93, in __del__
    warnings.warn(f"unclosed transport {self!r}", ResourceWarning,
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\proactor_events.py", line 57, in __repr__
    info.append(f'fd={self._sock.fileno()}')
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\windows_utils.py", line 102, in fileno
    raise ValueError("I/O operation on closed pipe")
ValueError: I/O operation on closed pipe
Exception ignored in: <function _ProactorBasePipeTransport.__del__ at 0x000001E59E9AD378>
Traceback (most recent call last):
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\proactor_events.py", line 93, in __del__
    warnings.warn(f"unclosed transport {self!r}", ResourceWarning,
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\proactor_events.py", line 57, in __repr__
    info.append(f'fd={self._sock.fileno()}')
  File "C:\Users\marc.otten\AppData\Local\Programs\Python\Python37\lib\asyncio\windows_utils.py", line 102, in fileno
    raise ValueError("I/O operation on closed pipe")
ValueError: I/O operation on closed pipe

```

**Comment by devtud** [comment]
It's pretty hard to follow what you've done there. Ideally you should only add 
```pyhton
if 'Finished composition' in log_line:
    process_task.kill()
```
inside `check_and_raise()` and retrieve errors in `asyncio.gather(...)`.

One mistake I see in your code is that you kill the process and then you read from its stdin - which probably causes the `I/O operation on closed pipe` value error.

**Comment by Marcuzzz** [comment]
Hey Devtud,

Just want to say thank you for pointing me into the right direction. I was unfortunately not able to get it to work.
I still not completely understand asyncio. 

I did write this code below that seems to work.
I've used a couple of your functions.

So again thank you!
Cheers

```
import subprocess
import threading
import logging
from pathlib import Path

logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger('aerender')

def output_reader(proc):

    for line in iter(proc.stdout.readline, b''):
        log_tag = 'stdout'
        line = line.decode('utf-8').strip()
        if 'PROGRESS:  Total Time Elapsed' in line:
            line = 'AE finished... exit code: ' + str(proc.returncode)
            log_tag = 'stdout'
            logger.info(f'{log_tag}: {line}')
            proc.terminate()
            return None

        logger.info(f'{log_tag}: {line}')


def get_command(
        aerender_abs_path: str, *, reuse: bool = None, project_path: str = None,
        project_name: str = None, comp_name: str = None,
        index_in_render_queue: str = None, render_settings_template: str = None,
        output_module_template: str = None, output_path: str = None,
        logfile_path: Path = None, start_frame: int = None, end_frame: int = None,
        increment: int = None, image_cache_percent: int = None,
        max_mem_percent: int = None, verbose_flag: str = None, close_flag: str = None,
        sound_flag: str = None, continue_on_missing_footage: bool = None,
        version: bool = None
    ) -> str:

    command = f'"{aerender_abs_path}"'
    if reuse:
        command += f' -reuse'
    if project_path is not None:
        command += f' -project "{project_path}"'
    if project_name is not None:
        command += f' -teamproject "{project_name}"'
    if comp_name is not None:
        command += f' -comp "{comp_name}"'
    if index_in_render_queue is not None:
        command += f' -rqindex {index_in_render_queue}'
    if render_settings_template is not None:
        command += f' -RStemplate "{render_settings_template}"'
    if output_module_template is not None:
        command += f' -OMtemplate "{output_module_template}"'
    if output_path is not None:
        command += f' -output "{output_path}"'
    if logfile_path is not None:
        command += f' -log "{logfile_path}"'
    if start_frame is not None:
        command += f' -s {start_frame}'
    if end_frame is not None:
        command += f' -e {end_frame}'
    if increment is not None:
        command += f' -i {increment}'
    if image_cache_percent is not None and max_mem_percent is not None:
        command += f' -mem_usage {image_cache_percent} {max_mem_percent}'
    if verbose_flag is not None:
        assert verbose_flag in ['ERRORS', 'ERRORS_AND_PROGRESS']
        command += f' -v {verbose_flag}'
    if close_flag is not None:
        assert close_flag in ['DO_NOT_SAVE_CHANGES', 'SAVE_CHANGES', 'DO_NOT_CLOSE']
        command += f' -close {close_flag}'
    if sound_flag is not None:
        assert sound_flag in ['ON', 'OFF']
        command += f' -sound {sound_flag}'
    if continue_on_missing_footage:
        command += f' -continueOnMissingFootage'
    if version:
        command += f' -version'
    return command


def aerender(output_path,aerender_path,project_path,comp):

    command = get_command(project_path=Path(project_path), comp_name=comp, output_path=Path(output_path),aerender_abs_path=Path(aerender_path))
    proc = subprocess.Popen(command,shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)

    t = threading.Thread(target=output_reader, args=(proc,))
    t.start()

    try:
        proc.wait()
        proc.terminate()
    except:
        line = 'subprocess ended earlier... '
        log_tag = ''             
        logger.info(f'{log_tag}: {line}')


    t.join()

#example input
aerender(output_path,aerender_path,project_path,comp)
```

**Comment by devtud** [comment]
Hey @Marcuzzz , I'm glad you found a solution that fits your needs!
